[EDU-618] Document Spot retry strategy (#5872)
nextflow-io/nextflow: docs/updating-spot-retries.md

(spot-retry-page)=

# Spot instance failures and retries

> This page describes changes in how Nextflow handles spot instance failures on AWS and Google Cloud, the impact of those changes, and how to configure spot retry behavior for your pipelines. It applies to Nextflow 24.10 and later.

Reviewer: ```suggestion
This page describes changes in how Nextflow handles spot instance failures on AWS and Google Cloud, the impact of those changes, and how to configure spot retry behavior for your pipelines. These changes apply to Nextflow 24.10 and later.
```

## Retry behavior

> Up to version 24.10, Nextflow silently retries spot instance failures up to `5` times when using AWS Batch or Google Batch. These retries were controlled by cloud-specific configuration parameters (e.g., `aws.batch.maxSpotAttempts`) and happened in cloud infrastructure without explicit visibility to Nextflow.

Reviewer: Since the rest of this paragraph is in past tense, it makes sense to apply the same to the first sentence. 

```suggestion
Up to version 24.10, Nextflow would silently retry spot instance failures up to `5` times when using AWS Batch or Google Batch. These retries were controlled by cloud-specific configuration parameters (e.g., `aws.batch.maxSpotAttempts`) and happened in cloud infrastructure without explicit visibility to Nextflow.
```

<h3>Before Nextflow 24.10</h3>

By default, Nextflow would instruct AWS and Google to automatically retry jobs lost to spot reclamation up to `5` times. Retries were handled by the cloud provider _within_ a Nextflow task. It was often unclear that tasks were restarted as there was no explicit message. Task runtimes and associated cloud costs were increased because they included the runtime of the reclaimed and retried tasks. Due to the high likelihood of reclamation before completion, long-running tasks running on spot instances frequently required retries, leading to inefficient allocation of resources and higher costs.

<h3>After Nextflow 24.10</h3>

> The default spot reclamation retry setting changed to `0` on AWS and Google. By default, no _internal_ retries are attempted on these platforms. Spot reclamations leads to an immediate failure that is exposed to Nextflow in the same way as other generic failures (for example, returning `exit code 1` on AWS). Nextflow treats these failures like any other job failure unless a retry strategy is configured.

Reviewer: ```suggestion
The default spot reclamation retry setting changed to `0` on AWS and Google. By default, no _internal_ retries are attempted on these platforms. Spot reclamations lead to an immediate failure, exposed to Nextflow in the same way as other generic failures (for example, returning `exit code 1` on AWS). Nextflow treats these failures like any other job failure, unless a retry strategy is configured.
```

> ## Impact on Existing Workflows

Reviewer: ```suggestion
## Impact on existing workflows
```

Changing here without looking at the rest of the NF docs, though we do standardize on sentence case headings in the Platform docs.

> If you have been relying on silent spot retries (the previous default behavior), you may now see more tasks fail with the following characteristics:

Reviewer: ```suggestion
If you rely on silent spot retries (the previous default behavior), you may now see more tasks fail with the following characteristics:
```

Since the default for spot retries is now `0`, you must actively enable a retry strategy if you want Nextflow to handle reclaimed spot instances automatically.

## Possible actions

There are four possible actions.

### Do nothing

If you do not configure anything, you will observe more pipeline failures when spot instances are reclaimed. This approach provides clearer visibility into failures. Failed tasks can be re-run with the `-resume` option. However, frequent task reclamation may lead to a higher failure rate and each retry requires manual intervention.

:::{note}
If you resume the pipeline using the resume option, it will pick up at the point the pipeline was interrupted and start with a retry of that task.
:::

### Re-enable spot retries

You can re-enable spot retries at the provider level in your Nextflow configuration:

```
// nextflow.config
aws {
    batch {
        maxSpotAttempts = 5
    }
}

google {
    batch {
        maxSpotAttempts = 5
    }
}
```

The above example sets the maximum number of spot retries to `5` for both AWS and Google.

> ### Make spot failures visible but retry them

Reviewer: Reads more positively? Ignore if this makes less sense. 

```suggestion
### Make spot failures visible and retry them
```

You can set `maxRetries` to enable Nextflow-level retries for any failure:

```
// nextflow.config
process {
    maxRetries = 5
}
```

The above example sets retries to `5` for any failures, not just failures at the provider level.

> ### Use Fusion snapshots

Reviewer: ```suggestion
### Use Fusion Snapshots (AWS Batch only)
```

> If you have long-running tasks where losing progress due to spot reclamation is costly, consider [Fusion snapshots](https://docs.seqera.io/fusion/guide/snapshots) (if supported by your environment). Fusion snapshots allows you to resume a partially completed task on a new machine if a spot instance is reclaimed and reduce wasted compute time.

Reviewer: ```suggestion
If you have long-running tasks where progress lost due to spot reclamation is costly, consider [Fusion Snapshots](https://docs.seqera.io/fusion/guide/snapshots) (if supported by your environment). Fusion Snapshots allow you to resume a partially completed task on a new machine if a spot instance is reclaimed, thereby reducing wasted compute time.
```

> Key features of Fusion snapshots:

Reviewer: ```suggestion
Key features of Fusion Snapshots:
```

- Jobs do not need to start from scratch after reclamation.
- Especially useful for tasks that take many hours or days to complete.
- May significantly reduce costs and run times in high-reclamation environments.

See [Fusion Snapshots for AWS Batch](https://docs.seqera.io/fusion/guide/snapshots) for more information.

## Best practices

Best practices for spot instance failures and retries:

> - **Evaluate job duration**: If your tasks are very long (for example, multi-hour, multi-day), spot instances can cause repeated interruptions. Consider using on-demand instances or Fusion snapshots.

Reviewer: ```suggestion
- **Evaluate job duration**: If your tasks are very long (multi-hour or multi-day), spot instances can cause repeated interruptions. Consider using on-demand instances or Fusion Snapshots.
```

- **Set sensible retry limits**: If you enable spot retries, choose a retry count that balances the cost savings of spot usage against the overhead of restarting tasks.

> - **Monitor logs and exit codes**: Failures due to spot reclamation will now appear in Nextflow logs. Monitoring failures and fine-tune your strategy.

Reviewer: ```suggestion
- **Monitor logs and exit codes**: Failures due to spot reclamation will now appear in Nextflow logs. Monitor failures and fine-tune your strategy.
```

> - **Consider partial usage of spot**: Some workflows may mix on-demand instances for critical or long tasks and spot instances for shorter less critical tasks. This can optimize cost while minimizing wasted compute time.

Reviewer: ```suggestion
- **Consider partial usage of spot**: Some workflows may mix on-demand instances for critical or long tasks and spot instances for shorter, less critical tasks. This can optimize cost while minimizing wasted compute time.
```
